### DISCUSSION TITLE
how can i optimize for ggml models running on cpu?

### DISCUSSION THREAD

**Discussion by yesbroc** [oobabooga/text-generation-webui#1600] [Q&A]
i have 8 cores 16 threads 2.9 ghz cpu with 16gb of ram, what do i enable/disable to get fast results like llama.cpp

**Comment by viperwasp** [comment]
Yes I asked similar question too. Even if I use the command to activate more threads it still only uses 18-22% of my CPU? I don't get it.

**Reply by yesbroc** [reply]
llama.cpp was basically faster in everything during my benchmarking, even if my cpu is bad it's still proof oobabooga isnt optimized for cpu

**Comment by BetaDoggo** [comment]
The webui uses the [llama-cpp-python](https://github.com/abetlen/llama-cpp-python)  bindings for its ggml support which is likely the source of the performance loss. The only performance related settings you can change are the number of threads(8 is ideal for your setup) and context length(minimum of 512).

**Comment by electric-samurai** [comment]
I too see this issue and have been investigating.  I don't buy that the issue is solely due to using a python wrapper for llama (simply because the intensive work is passed down to llama.cpp), but I do believe there may be a difference in how that wrapper sets up/uses the llama.cpp versus how it is used in llama.cpp alone.  One issue is that the wrapper (or I might be remembering incorrectly and its text-generation-webui), sets the number of threads by default to one half of your cpus.  You should specify the number of threads directly.  Still I'm not sure yet if this alone is enough to get the same performance.  One other possibility is that the wrapper is behind the main llama.cpp.  For that I have been building the wrapper from source and making sure its version of llama.cpp in its source is close to the current version.

**Comment by yesbroc** [comment]
i heard that 8 threads were the sweet spot, can someone confirm? ive defaulted to 12 threads bc double digits, Im not sure if higher=better here.

p.s ik this discussion is old but im trying to get close to 2.5 t/s, also using gpu accel but my gpu is actually dog (4gb vram)

**Reply by viperwasp** [reply]
I think the sweet spot depends on your given processor. I don't know all the science behind it. But I get optimal results with 8 threads on a 12700K. Maybe it's one thread per performance core? But that is only a guess?
